src/client/PyCallbacks.h

```diff
@@ -97,25 +101,44 @@ class PyCallbacks : public IPyCallbacks {
 		return mModels[initialShapeIdx].mCGAReport;
 	}
 
+	const std::wstring& getCGAPrints(const size_t initialShapeIdx) const {
+		if (initialShapeIdx >= mModels.size())
+			throw std::out_of_range("initial shape index is out of range.");
+
+		return mModels[initialShapeIdx].mCGAPrints;
+	}
+
+	const std::wstring& getCGAErrors(const size_t initialShapeIdx) const {
+		if (initialShapeIdx >= mModels.size())
+			throw std::out_of_range("initial shape index is out of range.");
+
+		return mModels[initialShapeIdx].mCGAErrors;
+	}
+
 	prt::Status generateError(size_t isIndex, prt::Status status, const wchar_t* message) override {
-		pybind11::print(L"GENERATE ERROR:", isIndex, status, message);
 		return prt::STATUS_OK;
 	}
 
 	prt::Status assetError(size_t isIndex, prt::CGAErrorLevel level, const wchar_t* key, const wchar_t* uri,
 	                       const wchar_t* message) override {
-		pybind11::print(L"ASSET ERROR:", isIndex, level, key, uri, message);
+		std::wstring errorMsg(L"Asset" + ERRORLEVELS[level] + key + L" " + uri + L" " + message);
+		mModels[isIndex].mCGAErrors += errorMsg;
+
 		return prt::STATUS_OK;
 	}
 
-	prt::Status cgaError(size_t isIndex, int32_t shapeID, prt::CGAErrorLevel level, int32_t methodId, int32_t pc,
+	prt::Status cgaError(size_t isIndex, int32_t /*shapeID*/, prt::CGAErrorLevel level, int32_t methodId, int32_t pc,
 	                     const wchar_t* message) override {
-		pybind11::print(L"CGA ERROR:", isIndex, shapeID, level, methodId, pc, message);
+		std::wstring errorMsg(L"CGA" + ERRORLEVELS[level] + message);
+		mModels[isIndex].mCGAErrors += errorMsg;
+
 		return prt::STATUS_OK;
 	}
 
-	prt::Status cgaPrint(size_t isIndex, int32_t shapeID, const wchar_t* txt) override {
-		pybind11::print(L"CGA PRINT:", isIndex, shapeID, txt);
+	prt::Status cgaPrint(size_t isIndex, int32_t /*shapeID*/, const wchar_t* txt) override {
+		std::wstring printsTxt(txt);
+		mModels[isIndex].mCGAPrints += printsTxt;

```

Comment:
Also ensure newline...

Reply:
I don't think it's a good idea in this case. The reason is that it will not reflect what the cga code does/should do.
Moreover, we wouldn't have the same behavior as in CityEngine.